interfaces: PlugInfo/SlotInfo/ConnectedPlug/ConnectedSlot attribute helpers

stolowski opened the pull request:
Small helpers to ease with interfaces refactoring:
- Attr() getter for PlugInfo and SlotInfo that reflect ConnectdPlug/ConnectedSlot API. Once the transition is completed, the Attrs can be changed to private.
- AttrGettr interface to simplify change to some helper methods in interfaces (e.g. in content or dbus) where  they need to read attributes in sanitize methods and connected plug/slot methods; with this simple interface they can read attr from PlugInfo/SlotInfo/ConnectedSlot/ConnectedPlug.
- Updated ConnectedPlug/ConnectedSlot Attr getters: applied Maciej's suggestion about using reflection.

zyga commented on snap/info.go:
```diff
@@ -415,6 +422,13 @@ func (plug *PlugInfo) String() string {
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
```
It would be great if you could just test this explicitly in a separate test method.

bboozzoo commented on snap/info.go:
```diff
@@ -397,6 +397,13 @@ type PlugInfo struct {
 	Hooks     map[string]*HookInfo
 }
 
+func (plug *PlugInfo) Attr(key string) (interface{}, error) {
```
I guess the caller will do a type assert because this is returning `interface{}`. In such case you may refactor this into
```
import "reflect"
...
func (plug *PlugInfo) Attr(key string, val interface{}) error {
	rt := reflect.TypeOf(val)
	if rt.Kind() != reflect.Ptr || val == nil {
		return errors.New("not assignable")
	}
	if v, ok := plug.Attrs[name]; ok {
		if reflect.TypeOf(v) != rt.Elem() {
			return errors.New("type mismatch")
		}
		rv := reflect.ValueOf(val)
		rv.Elem().Set(reflect.ValueOf(v))
		return nil
	}
	return errors.New("not found")
}
```

Then the call sites will be cleaner:

```
var foo string
if err := plug.Attr("foo", &foo); err != nil {
    // blah blah failed
}
```

A working example is right here: https://play.golang.org/p/tyvQ0RpeFA

stolowski replied:
Thanks, I like that, it would simplify the code of interfaces a little bit. @pedronis , @niemeyer your opinion on that? If that's ok I'll update the ConnectedPlug/ConnectedSlot to match.

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
 
+// AttrGetter is an interface with Attr getter method common
+// to ConnectedSlot, ConnectedPlug, PlugInfo and SlotInfo types.
+type AttrGetter interface {
```
This should probably be Attrer or similar. Given typical conventions, it feels like an AttrGet method should exist.

That said, this looks like a left over. There's no usage of the interface in the whole PR, so I suggest just dropping it.

stolowski replied:
_Attrer_, did you actually mean that? How about _AttrReader_?

The _Attr_ method name needs to match the method of ConnectedPlug/Slot and PlugInfo/SlotInfo.
This interface is used in the followup branch, I just forgot to update it to match the changes to _Attr_ methods. Fixed now.

pedronis replied:
I think Gustavo did meant Attrer, I can see at least an example where in the stdlib they went for such blunt/ungrammatical naming  https://golang.org/pkg/database/sql/driver/#Valuer

stolowski replied:
I see. Ok, renamed.

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
+	var v interface{}
+	var ok bool
+
+	if dynamicAttrs != nil {
```
The three lines above can go away and be replaced with just:
```
v, ok := dynamicAttrs[key]
```
This will work as you want if the map is nil.

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
+	if dynamicAttrs != nil {
+		v, ok = dynamicAttrs[key]
+	}
+	if !ok && staticAttrs != nil {
```
For the same reason, this can be simply:
```
if !ok {
```

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
+		v, ok = staticAttrs[key]
+	}
+
+	if ok {
```
Instead of indenting everything and having to keep in mind what the condition was by the time we get to the last line, it's cleaner to invert the logic and repeat the above statement:
```
if !ok {
    return fmt.Errorf("attribute %q not found", key)
}
(... rest of the logic ...)
```

niemeyer replied:
This would benefit from context as well. See the related review point.

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
+	if ok {
+		rt := reflect.TypeOf(val)
+		if rt.Kind() != reflect.Ptr || val == nil {
+			return fmt.Errorf("cannot store the value of attribute %q", key)
```
This error message does not reflect the action being done. Neither the developer nor the user at the other end have asked to store an attribute. Should be something along the lines of:

_"internal error: cannot get attribute with non-pointer value"_

niemeyer replied:
Similarly, we'd benefit from some context. Which interface type, which attribute. Otherwise we'll need to blindly hunt for the error when someone reports this.

niemeyer commented on interfaces/connection.go:
```diff
@@ -45,6 +46,39 @@ type ConnectedSlot struct {
+		}
+
+		if reflect.TypeOf(v) != rt.Elem() {
+			return fmt.Errorf("the type of attribute %q is %s, expected %s", key, reflect.TypeOf(v).Name(), rt)
```
Similarly, this message is going to surface to users due to errors from snap publishers. Neither of those need to be Go developers, so we don't want to expose strings such as _*map[string]interface{}_ in an error message.

Also, this is lacking some context that would enable to user or publisher to tell _where_ that attribute is.

Needs to be something similar to:

_"snap %q has interface %q with invalid value type for %q attribute"_

If you tweak the message further, let's please just preserve the hierarchy in the order of the arguments: snap > interface > attribute.

stolowski commented:
@niemeyer Thanks for the review, I've addressed your points re error messages and code simplifications. Except for `AttrGetter` - please see my comment.

pedronis commented on interfaces/connection.go:
```diff
@@ -49,6 +50,38 @@ type ConnectedSlot struct {
+	}
+
+	if reflect.TypeOf(v) != rt.Elem() {
+		return fmt.Errorf("snap %q has interface %q with invalid value type for %q attribute", snapName, ifaceName, key)
```
I find this error a bit too abstract,

"snap %q has interface %q with %q attribute value of invalid type %v, expected %v", snapName, ifaceName, key, reflect.TypeOf(v), rt.Elem()

stolowski replied:
I included type names initially but changed this per earlier request from Gustavo: 

"Similarly, this message is going to surface to users due to errors from snap publishers. Neither of those need to be Go developers, so we don't want to expose strings such as *map[string]interface{} in an error message.

Also, this is lacking some context that would enable to user or publisher to tell where that attribute is.

Needs to be something similar to:

"snap %q has interface %q with invalid value type for %q attribute"

pedronis replied:
but 'invalid value type' is not very user friendly either, anyway my worry is that we will miss debug info, maybe we should log the message with types, and surface the one without:

"snap %q has interface %q with %q attribute value with unexpected type"

stolowski replied:
Fair enough, I'll address this in a follow up PR. Thanks.

pedronis commented on snap/info.go:
```diff
@@ -398,6 +399,27 @@ type PlugInfo struct {
+		}
+
+		if reflect.TypeOf(v) != rt.Elem() {
+			return fmt.Errorf("snap %q has interface %q with invalid value type for %q attribute", snapName, ifaceName, key)
```
same wondering about the error
